gateway/controller.go

```diff
@@ -18,90 +22,128 @@ const (
 type ControllerClient interface {
 	GetSchemaHash(namespace string) (string, error)
 	GetAllSchema(namespace string) ([]common.Table, error)
+	GetAssignmentHash(jobNamespace, instance string) (string, error)
+	GetAssignment(jobNamespace, instance string) (*rules.Assignment, error)
 }
 
 // ControllerHTTPClient implements ControllerClient over http
 type ControllerHTTPClient struct {
-	c              *http.Client
-	controllerHost string
-	controllerPort int
-	headers        http.Header
+	c       *http.Client
+	address string
+	headers http.Header
 }
 
 // NewControllerHTTPClient returns new ControllerHTTPClient
-func NewControllerHTTPClient(controllerHost string, controllerPort int, headers http.Header) *ControllerHTTPClient {
+func NewControllerHTTPClient(address string, timeoutSec time.Duration, headers http.Header) *ControllerHTTPClient {
 	return &ControllerHTTPClient{
-		c:              &http.Client{},
-		controllerHost: controllerHost,
-		controllerPort: controllerPort,
-		headers:        headers,
+		c: &http.Client{
+			Timeout: timeoutSec,
+		},
+		address: address,
+		headers: headers,
 	}
 }
 
-func (c *ControllerHTTPClient) GetSchemaHash(namespace string) (hash string, err error) {
-	var req *http.Request
-	req, err = c.getRequest(namespace, true)
+// buildRequest builds an http.Request with headers.
+func (c *ControllerHTTPClient) buildRequest(method, path string) (req *http.Request, err error) {
+	path = strings.TrimPrefix(path, "/")
+	url := fmt.Sprintf("http://%s/%s", c.address, path)
+	req, err = http.NewRequest(method, url, nil)
 	if err != nil {
-		return
+		return nil, err
 	}
-	var resp *http.Response
-	resp, err = c.c.Do(req)
+
+	req.Header = c.headers
+	req.Header.Add("RPC-Procedure", path)
+	return
+}
+
+func (c *ControllerHTTPClient) getResponse(request *http.Request) ([]byte, error) {
```

Comment:
nit: let's be consistent to use named return type.

Reply:
done